src/main/java/core/stack/MateStack.java

```diff
@@ -1,19 +1,48 @@
 package core.stack;
 
+import java.util.EmptyStackException;
+
 public class MateStack<T> {
-    public void push(T value) {
+    private Node<T> top;
+    private int size;
+
+    private static class Node<T> {
+        private final T value;
+        private final Node<T> previous;
 
+        public Node(T value, Node<T> previous) {
+            this.value = value;
+            this.previous = previous;
+        }
+    }
+
+    public void push(T value) {
+        if (size == 0) {
+            top = new Node<>(value, null);
+        } else {
+            top = new Node<>(value, top);
+        }
```

Comment:
Initially when the stack is empty the Top field is null. you don't need this check here.
```suggestion
            top = new Node<>(value, top);
```